### PR TITLE
Add support for backward_passes_per_step > 1 for LegacyOptimizers (TF) in Graph Mode.

### PR DESCRIPTION
**Pull request by aaron276h**
## Checklist before submitting

- [x] Did you read the [contributor guide](https://github.com/horovod/horovod/blob/master/CONTRIBUTING.md)?
- [ ] Did you update the docs?
- [x] Did you write any tests to validate this change?  
- [x] Did you update the [CHANGELOG](https://github.com/horovod/horovod/blob/master/CHANGELOG.md), if this change affects users?

## Description
This PR is a follow up #2346. This PR adds support for backward_passes_per_step > 1 for TF legacy optimizers (tf.train.Optimizer) executing in graph(non-eager) mode. This is one of the features that we have built into[ Determined AI's](https://determined.ai/) fork of Horovod that we would like to upstream.

## Review process to land 

1. All tests and other checks must succeed.
2. At least one member of the [technical steering committee](https://github.com/horovod/horovod/blob/master/CONTRIBUTING.md) must review and approve.
3. If any member of the technical steering committee requests changes, they must be addressed.

### CODE REVIEW

**Review by tgaddair — APPROVED** [review]
Nice work!  Awesome to see feature parity among the different optimizers.

**Comment by Richie-yan** [comment]
Hi, @aaron276h 
When I ran tensorflow_mnist_estimator.py using gradient accumulation, I set the backward_passes_per_step to 3 and encountered the following problems:
屏幕快照 2020-11-02 上午11 55 29
When the broadcast is shown in the above picture, the counter names of the two ranks are inconsistent, which leads to hang. I think it is related to the following code:
屏幕快照 2020-11-02 下午12 06 48
Does this code need to distinguish rank on the counter variable?

**Comment by aaron276h** [comment]
@Richie-yan thanks for flagging this issue. Could you take a look at #2415, that should address the issue you are running into.

**Comment by Richie-yan** [comment]
Hi, @aaron276h @tgaddair 
I tried to run TF's Roberta large model using gradient accumulation, and found a problem:
Set batch_size to 10：

- Do not use gradient accumulation, TF's bfc memory usage is about 23.3GB;

- Using gradient accumulation, TF's bfc memory usage is about 29.2GB.

According to my statistics, the gradient of the Roberta large model occupies about 1.17GB of GPU memory.
However, using gradient accumulation will take up an additional 5.9GB of GPU memory. 
When using gradient accumulation, the use of GPU memory does not meet expectations. What may be the cause?

**Comment by tgaddair** [comment]
Good catch @Richie-yan.  I'm not sure it's possible to avoid this, as the gradients need to be stored into a separate variable in order to perform the local aggregation.  Do you have some thoughts on how this additional copy can be avoided?

**Comment by Richie-yan** [comment]
@tgaddair 
thank you for your reply
I haven't thought of other methods that can replace this kind of copy.
However, I tried to modify the variable initialization method, and it seems that the use of video memory has returned to normal, from the original additional 5.9GB to about 1.3GB. My method is as follows:
屏幕快照 2020-11-17 上午10 22 48

**Comment by aaron276h** [comment]
@Richie-yan that's really interesting, this should give the correct performance and if you are observing that this provides better memory performance we should definitely make this change. 

Seems to be potentially related to this old thread but it's not clear why we see 5x memory usage rather than 2x from using `tf.zeroes_initializer()`.

**Comment by Richie-yan** [comment]
@aaron276h 
Yes, changing to the above method, I can further increase the batch_size of the model without OOM, thereby improving the throughput of the model.
I think we can further investigate why 5x memory usage is occupied using `tf.zeroes_initializer().`
